File: ios/MoneyMaskEngine.swift
import NitroModules

struct MoneyMaskEngine: MaskEngineProtocol {
  private let precision: Int
  private let separator: String
  private let delimiter: String
  private let unit: String
  private let suffixUnit: String
  private let zeroCents: Bool

  init(options: NitroMaskOptions) {
    self.precision   = Int(options.precision ?? 2)
    self.separator   = options.separator  ?? ","
    self.delimiter   = options.delimiter  ?? "."
    self.unit        = options.unit       ?? ""
    self.suffixUnit  = options.suffixUnit ?? ""
    self.zeroCents   = options.zeroCents  ?? false
  }

  func apply(input: String) -> (masked: String, raw: String) {
    let digits = input.filter { $0.isNumber }

    if digits.isEmpty {
      return ("", "")
    }

    let effectivePrecision = zeroCents ? 0 : precision

    // Pad on the left so we have at least (precision+1) digits
    let minLength = effectivePrecision + 1
    let padded = String(repeating: "0", count: max(0, minLength - digits.count)) + digits

    // Split integer and decimal parts
    let splitIdx: String.Index
    if effectivePrecision > 0 {
      splitIdx = padded.index(padded.endIndex, offsetBy: -effectivePrecision)
    } else {
      splitIdx = padded.endIndex
    }

    var intPart = String(padded[padded.startIndex..<splitIdx])
    let centPart = effectivePrecision > 0 ? String(padded[splitIdx...]) : ""

    // Remove leading zeros from intPart, keep at least "0"
    while intPart.count > 1 && intPart.hasPrefix("0") {
      intPart = String(intPart.dropFirst())
    }

    // Insert delimiter every 3 digits from right
    var formattedInt = ""
    let intChars = Array(intPart)
    let total = intChars.count
    for (i, c) in intChars.enumerated() {
      formattedInt.append(c)
      let remaining = total - i - 1
      if remaining > 0 && remaining % 3 == 0 {
        formattedInt += delimiter
      }
    }

    // Build result
    var masked = unit + formattedInt
    if effectivePrecision > 0 {
      masked += separator + centPart
    }
    masked += suffixUnit

    return (masked, digits)
  }

  var wantsTrailingCursor: Bool { true }
  var expandedMask: String? { nil }
}

File: android/src/main/java/com/nitroinputmask/MoneyMaskEngine.kt
package com.nitroinputmask

import com.margelo.nitro.nitroinputmask.NitroMaskOptions

internal class MoneyMaskEngine(options: NitroMaskOptions) : MaskEngineProtocol {
  private val precision: Int = (options.precision?.toInt()) ?: 2
  private val separator: String = options.separator ?: ","
  private val delimiter: String = options.delimiter ?: "."
  private val unit: String = options.unit ?: ""
  private val suffixUnit: String = options.suffixUnit ?: ""
  private val zeroCents: Boolean = options.zeroCents ?: false

  override fun apply(input: String): Pair<String, String> {
    val digits = input.filter { it.isDigit() }

    if (digits.isEmpty()) {
      return Pair("", "")
    }

    val effectivePrecision = if (zeroCents) 0 else precision

    // Pad on the left so we have at least (precision+1) digits
    val minLength = effectivePrecision + 1
    val padded = digits.padStart(maxOf(minLength, digits.length), '0')

    // Split integer and decimal parts
    var intPart: String
    val centPart: String
    if (effectivePrecision > 0) {
      intPart = padded.dropLast(effectivePrecision)
      centPart = padded.takeLast(effectivePrecision)
    } else {
      intPart = padded
      centPart = ""
    }

    // Remove leading zeros from intPart, keep at least "0"
    while (intPart.length > 1 && intPart.startsWith('0')) {
      intPart = intPart.drop(1)
    }

    // Insert delimiter every 3 digits from right
    val sb = StringBuilder()
    val total = intPart.length
    for (i in intPart.indices) {
      sb.append(intPart[i])
      val remaining = total - i - 1
      if (remaining > 0 && remaining % 3 == 0) {
        sb.append(delimiter)
      }
    }

    // Build result
    var masked = unit + sb.toString()
    if (effectivePrecision > 0) {
      masked += separator + centPart
    }
    masked += suffixUnit

    return Pair(masked, digits)
  }

  override val wantsTrailingCursor: Boolean = true
  override val expandedMask: String? = null
}
